Neo4j.Driver/Neo4j.Driver/Internal/Routing/AddressSet.cs

```diff
@@ -14,136 +14,105 @@
     {
-        private readonly IList<T> _items = new List<T>();
-        private int _index = 0;
-
-        internal int Index => _index;
+        private readonly object _itemsLock = new object();
+        private volatile IList<T> _items = new List<T>();
 
         /// <summary>
-        /// Add one item into this set.
+        ///     Add one item into this set.
         /// </summary>
         /// <param name="item">The item to add</param>
         public void Add(T item)
         {
-            lock (_items)
+            lock (_itemsLock)
             {
                 if (!_items.Contains(item))
                 {
-                    _items.Add(item);
+                    var newItems = new List<T>(_items) {item};
+                    _items = newItems;
                 }
             }
         }
 
         /// <summary>
-        /// Adds several _items into this set.
+        ///     Adds several _items into this set.
         /// </summary>
         /// <param name="items">The _items to add</param>
         public void Add(IEnumerable<T> items)
         {
-            lock (_items)
+            lock (_itemsLock)
             {
+                var newItems = new List<T>(_items);
                 foreach (var item in items)
                 {
-                    Add(item);
+                    if (!newItems.Contains(item))
+                    {
+                        newItems.Add(item);
+                    }
                 }
+                _items = newItems;
             }
         }
 
         /// <summary>
-        /// Remove one item from this set
+        ///     Remove one item from this set
         /// </summary>
         /// <param name="item"></param>
         public void Remove(T item)
         {
-            lock (_items)
+            lock (_itemsLock)
             {
-                var pos = _items.IndexOf(item);
-                _items.Remove(item);
-                if (_index > pos)
-                {
-                    _index--;
-                }
+                var newItems = new List<T>(_items);
+                newItems.Remove(item);
+                _items = newItems;
             }
         }
 
         /// <summary>
-        /// Round robin to get the next item in the set
+        ///     Clean all the item inside this set
         /// </summary>
-        /// <param name="value">The next item in the set</param>
-        /// <returns>true if succesfully find an item, otherwise false if the set is empty</returns>
-        public bool TryNext(out T value)
+        public void Clear()
         {
-            lock (_items)
-            {
-                if (_items.Count == 0)
-                {
-                    value = default(T);
-                    return false;
-                }
-                // ensure the index is in range
-                _index = _index % _items.Count;
-                value = _items[_index++];
-            }
-            return true;
+            _items = new List<T>();
```

Comment:
Yes, good point. It is fine for `#Clear()` to assign because it is an atomic action, but it is not fine for it to not coordinate with other modification methods like `#Add()` or `#Remove()`.